adding accurate point turn functionality

mvoodarla opened the pull request:
Make comments here!

Commit:
initial point turn commit

mvoodarla commented on src/org/usfirst/frc/team2473/robot/commands/PointTurn.java:
```diff
@@ -0,0 +1,94 @@
+	protected void execute() {
+		double currDegrees = Devices.getInstance().getNavXGyro().getAngle();
+		double delta = currDegrees - prevDegrees;
+		boolean slow = degreesGoal - (currDegrees + delta) < RobotMap.K_DEGREE_THRESHOLD;
```
`(degreesGoal - (currDegrees + delta)` can be negative when turning clockwise
```suggestion
		boolean slow = Math.abs(degreesGoal - (currDegrees + delta) < RobotMap.K_DEGREE_THRESHOLD);
```

mvoodarla committed:
Fixed issue with slow check being negative

mvoodarla committed:
point turn tested but still innacurate by a factor for every angle

mvoodarla commented on src/org/usfirst/frc/team2473/robot/commands/PointTurn.java:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	@Override
+	protected void execute() {
```
Add a gradual dampening of motor power for more accurate turning

mvoodarla commented on src/org/usfirst/frc/team2473/framework/Database.java:
```diff
@@ -0,0 +1,99 @@
+ * 
+ *
+ */
+public class Database {
```
Not using this class anywhere, can delete.

mvoodarla commented on src/org/usfirst/frc/team2473/framework/RobotTalon.java:
```diff
@@ -0,0 +1,21 @@
+	
+	@Override
+	public void set(double pow) {
+		super.set(pow*PREFERRED_VOLTAGE/RobotController.getBatteryVoltage());
```
Can remove this class too, we can think about implementing this type of adjustment for low voltage problems later.

mvoodarla commented on src/org/usfirst/frc/team2473/framework/TrackableSubsystem.java:
```diff
@@ -0,0 +1,12 @@
+
+import edu.wpi.first.wpilibj.command.Subsystem;
+
+public abstract class TrackableSubsystem extends Subsystem {
```
Remove this class.

mvoodarla commented on src/org/usfirst/frc/team2473/robot/commands/DriveForDistanceCommand.java:
```diff
@@ -0,0 +1,88 @@
+
+import edu.wpi.first.wpilibj.command.Command;
+
+public class DriveForDistanceCommand extends Command {
```
Rename to `DriveStraight`

mvoodarla commented on src/org/usfirst/frc/team2473/robot/commands/PointTurn.java:
```diff
@@ -0,0 +1,110 @@
+
+public class PointTurn extends Command {
+	
+	private static double SLOW_POWER = 0.06;
```
Make this a constant in RobotMap

Commit:
removed unecessary classes and renamed drive straight

Commit:
changed PointTurn execute to dampen angle after a degree threshold

iamef commented on src/org/usfirst/frc/team2473/robot/commands/PointTurn.java:
```diff
@@ -16,91 +16,73 @@
 
 public class PointTurn extends Command {
 	
-	private static double SLOW_POWER = 0.06;
```
Slow power was deleted porque no es necesario para el dampening

iamef commented on src/org/usfirst/frc/team2473/robot/commands/PointTurn.java:
```diff
@@ -16,91 +16,73 @@
+		this.rightPower = -leftPower;	
 	}
-
+	
```
make a javadoc comment here to let the user know that is will always be positive

tnarayanan committed:
Fixed gradual slowdown, need to fix overshoot after robot stop

iamef committed:
modified linearization of power calibration, need to test

mvoodarla committed:
added further dampening to point turn at angle closer to goal

mvoodarla committed:
initial drive code commit

mvoodarla committed:
changed camera variable names

Co-authored-by: Tejas Narayanan <tejasn100@gmail.com>

tnarayanan committed:
Merge pull request #3 from Tino-FRC-2473/BasicDriveCode

basic drive code functionality

mvoodarla committed:
Stall experiment changes, K_TURN change to PointTurn

mvoodarla committed:
Working 90 degree turn with negative slowdown

mvoodarla committed:
Remove the build directory from git

mvoodarla committed:
update gitignore

tnarayanan committed:
Successful command group testing with point turn and drive straight working pretty well

mvoodarla committed:
new and improved gitignore

mvoodarla committed:
removed ignored files

mvoodarla committed:
Reset encoders in auto init

Earlier, we were resetting encoders in robot init which caused resetting
problems in drive straight but now it is fixed.

mvoodarla committed:
Temporary commit

My battery is low

Commit:
Ensure sign of encoder ticks is consistent

Before, our front right encoders were reporting negative even
though the robot was moving forward. This caused error in drive
straight and turning. Change getEncoderTicks method to reverse
sign of right side.

tnarayanan committed:
Fix StraightDrive for negative distances

StraightDrive has been updated to accomodate driving backwards.
Additionally, a temporary fix for encoder counts has been added to wait
until the encoder value reaches zero before resetting the next encoder.
This was added because encoder values were still getting reset in the
middle of the StraightDrive command, despite a delay of 5000ms in the
reset encoder function.

tnarayanan commented on src/org/usfirst/frc/team2473/robot/commands/StraightDrive.java:
```diff
@@ -29,7 +29,7 @@ public StraightDrive(double inches, double power) {
 		
 		this.inches = inches;
-		this.power = power;
+		this.power = (inches < 0) ? -power : power;
```
If both `inches` and `power` are negative, the power will be reversed and will be positive. An additional check needs to be made.

tnarayanan replied:
An alternative solution to this is to throw an `IllegalArgumentException` if the user inputs a `power` less than 0.

tnarayanan commented on src/org/usfirst/frc/team2473/robot/subsystems/DriveSubsystem.java:
```diff
@@ -152,21 +152,29 @@ public int getEncoderTicks(int id) {
 			throw new IllegalArgumentException(c.toString());
 		}
+		
+		while (motor.getSelectedSensorPosition() != 0);
```
This has not been tested yet. It was added because the 5000ms timeout on line 168 did not appear to fix the issue of encoders resetting in the middle of the StraightDrive command.

tnarayanan commented on src/org/usfirst/frc/team2473/robot/commands/PointTurn.java:
```diff
@@ -0,0 +1,126 @@
+		
+		if (degreesToGoal <= 10 && movingInTurnDirection) {
+			if (isClockwise) {
+				Robot.driveSubsystem.driveRawPower(-RobotMap.K_OPPOSITE_POWER, -RobotMap.K_OPPOSITE_POWER, RobotMap.K_OPPOSITE_POWER, RobotMap.K_OPPOSITE_POWER);
```
Using `driveRawPower()` could result in the robot shifting position during turns because the motors are not actually spinning at the same rate. We can try using `drive()` so that the lookup tables will balance the motor values and the robot will turn in place.

michelllepan committed:
Change encoder resets to asynchronous

Previously we used a 5000 ms timeout in setSelectedSensorPosition(). We
removed this timeout and made our resetEncoders() method synchronized so
that the robot intialization will wait for the encoders to
reset.

mvoodarla commented:
Not using this pull request anymore. See [#8](https://github.com/Tino-FRC-2473/FRC2018-Offseason/pull/8).
